src/features/RecipeEdit/RecipeEditController.tsx

```diff
@@ -1,20 +1,89 @@
-import { withRouter } from "react-router-dom";
-import useIngredientLO from "data/useIngredientLO";
+import { useGetAllLabels } from "data/hooks/useGetAllLabels";
+import { Redirect, useHistory, withRouter } from "react-router-dom";
 import { RouteComponentProps } from "react-router";
-import LoadObject from "../../util/LoadObject";
-import { Recipe } from "features/RecipeEdit/types";
+import { useGetRecipe } from "data/hooks/useGetRecipe";
+import PageBody from "views/common/PageBody";
+import RecipeForm from "features/RecipeEdit/components/RecipeForm";
+import DeleteButton from "views/common/DeleteButton";
+import RecipeApi from "data/RecipeApi";
+import { CircularProgress } from "@mui/material";
+import { useUpdateRecipe } from "data/hooks/useUpdateRecipe";
+import { useCreateRecipe } from "data/hooks/useCreateRecipe";
+import type { BfsId } from "global/types/identity";
+import type { DraftRecipe } from "global/types/types";
 
 type Props = RouteComponentProps<{ id?: string }>;
 
 const RecipeEditController: React.FC<Props> = ({ match }) => {
     const id = match.params?.id || "";
-    const lo = useIngredientLO(parseInt(id, 10));
+    const history = useHistory();
+    const { loading, error, data: recipe } = useGetRecipe(id);
     const { data: labelList } = useGetAllLabels();
+    const { updateRecipe } = useUpdateRecipe();
+    const { createRecipe } = useCreateRecipe();
+    let draft;
+
+    if (!id) {
+        return <Redirect to="/library" />;
+    }
+
+    if (loading || !recipe) {
+        return <CircularProgress />;
+    }
+
+    if (error) {
+        // TODO -- error message
+    }
+
+    const shouldCreateCopy = match.path.split("/").includes("make-copy");
+
+    draft = { ...recipe };
+    if (shouldCreateCopy) {
+        draft = { ...recipe, name: `Copy of ${recipe.name}` };
+    }
+
+    const handleUpdate = (recipe: DraftRecipe) => {
+        updateRecipe(recipe).then((_) => {
+            history.push(`/library/recipe/${recipe.id}`);
+        });
+    };
+
+    const handleSaveCopy = (recipe: DraftRecipe) => {
+        createRecipe(recipe).then((result) => {
+            const id = result.data?.library?.createRecipe.id;
+            id
+                ? history.push(`/library/recipe/${id}`)
+                : history.push(`/library`);
+        });
+    };
+
+    const handleDelete = (id: BfsId) => {
+        RecipeApi.deleteRecipe(id);
+    };
+
+    const handleCancel = () => {
+        history.push(`/library/recipe/${id}`);
+    };
+
     return (
-        <RecipeEdit recipeLO={lo as LoadObject<Recipe>} labelList={labelList} />
+        <PageBody>
+            <RecipeForm
+                title={`Editing ${draft.name}`}
+                recipe={draft}
+                onSave={shouldCreateCopy ? handleSaveCopy : handleUpdate}
+                onSaveCopy={handleSaveCopy}
+                onCancel={handleCancel}
+                labelList={labelList}
+                extraButtons={
+                    <DeleteButton
```

Comment:
I abstracted `DeleteButton` at the wrong layer, didn't I? Used an implementation where I should have used data and an agnostic renderer. This guy, sheesh.